lib/Dialect/TensorExt/Transforms/SIMDVectorize.cpp

```diff
@@ -0,0 +1,210 @@
+      if (failed(rewriter.convertRegionTypes(newRegion, *this->typeConverter)))
+        return failure();
+      regions.emplace_back(newRegion);
+    }
+
+    Operation *newOp = rewriter.create(OperationState(
+        op->getLoc(), op->getName().getStringRef(), operands, newResultTypes,
+        op->getAttrs(), op->getSuccessors(), regions));
+
+    rewriter.replaceOp(op, newOp);
+    return success();
+  }
+};
+
+struct ConvertGenericOp : public OpConversionPattern<secret::GenericOp> {
+  using OpConversionPattern::OpConversionPattern;
+
+  ConvertGenericOp(MLIRContext *context)
+      : OpConversionPattern<secret::GenericOp>(context) {}
+
+  LogicalResult matchAndRewrite(
+      secret::GenericOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    llvm::SmallVector<mlir::Type> newResultTypes;
+    if (failed(typeConverter->convertTypes(op->getResultTypes(),
+                                           newResultTypes))) {
+      return mlir::failure();
+    }
+
+    auto newOp = rewriter.create<secret::GenericOp>(
+        op->getLoc(), adaptor.getOperands(), newResultTypes,
+        [&](OpBuilder &b, Location loc, ValueRange blockArguments) {
+          IRMapping mp;
+          for (BlockArgument blockArg : op.getBody()->getArguments()) {
+            mp.map(blockArg, blockArguments[blockArg.getArgNumber()]);
+          }
+          for (auto &op : op.getBody()->getOperations()) {
+            b.clone(op, mp);
+          }
+        });
+    rewriter.replaceOp(op, newOp);
+    return success();
+  }
+};
+
+struct SIMDVectorize : impl::SIMDVectorizeBase<SIMDVectorize> {
+  using SIMDVectorizeBase::SIMDVectorizeBase;
+
+  void runOnOperation() override {
+    // Promote secret tensor types of arbitrary dimension into one specified by
+    // numSlots.
+    // Open questions: should this only run on secret tensor types? For
+    // plaintext matrices, we likely need to pack into a ciphertext if it is
+    // cheaper than many MulPlains. (e.g. a dot product with a plaintext weight
+    // matrix).
+    MLIRContext *context = &getContext();
+    auto *module = getOperation();
+
+    RewritePatternSet patterns(context);
+    ConversionTarget target(*context);
+    SecretTensorTypeConverter typeConverter(numSlots);
+    target.addLegalOp<ModuleOp>();
+    target.markUnknownOpDynamicallyLegal([&](Operation *op) {
+      op->dump();
+      return typeConverter.isLegal(op);
+    });
+
+    /* Require these since markUnknownOpDynamicallyLegal doesn't seem to work.
+    target.addDynamicallyLegalOp<secret::GenericOp>(
+        [&](secret::GenericOp op) { return typeConverter.isLegal(op); });
+
+    target.addDynamicallyLegalOp<secret::YieldOp>(
+        [&](secret::YieldOp op) { return typeConverter.isLegal(op); });
+
+    target.addDynamicallyLegalOp<arith::AddIOp>(
+        [&](arith::AddIOp op) { return typeConverter.isLegal(op); });
+    */
```

Comment:
It doesn't seem like `markUnknownOpDynamicallyLegal` is called to trigger on any ops

Reply:
How's it going with this? Anything I can do to assist?

Reply:
Right now the blocker is this comment :/ I think I'm missing some subtlety with the expected usage of `markUnknownOpDynamicallyLegal`, since it doesn't seem to be called at all - let me ask in MLIR discord

I'm not sure, you might be right?

Reply:
I think you might just want `addDynamicallyLegalOp`? https://mlir.llvm.org/doxygen/classmlir_1_1ConversionTarget.html#a6431de36d0202cc955a09372ea0e5303

Reply:
> I'm not sure, you might be right?

Upon reading the ~documentation~ code, I think your interpretation is the right one, so not sure why it's not being called.

Reply:
Let me play around with this, I lost context since last week so maybe my fresh eyes will help

Reply:
OK, I remember that `applyFullConversion` was one change I made that was necessary but still not sufficient weirdly. https://android.googlesource.com/platform/external/tensorflow/+/6b511124eb0fd1fdee83d25be9b5a1d10a31abe0/tensorflow/compiler/mlir/xla/transforms/legalize_tf_types.cc seems to be exactly similar to what I'm doing here, so trying to diff what's going on

Reply:
I found the problem! `addStructuralConversionPatterns` already includes a callback for `markUnknownOpDynamicallyLegal`, so it's overriding the one that I had set

Reply:
fixed!

Reply:
> fixed!

Awesome! Is it enough to have `markUnknownOpDynamicallyLegal` after `addStructuralConversionPatterns`?

Reply:
Yeah, although it's a bit of a "hack" so I'll add a comment on it - really we want to compose the legality callback with any existing unknown op legality callback, but there's no way to do that